### PR TITLE
walk the DOM and rot13 text nodes in place

### PR DESCRIPTION
**Pull request by wisnij**
This change preserves the structure of the original page and rot13's the text in place, rather than creating a single text node for the entire selection.

### CODE REVIEW

File: replace_text.js [hunk 1]
```diff
@@ -1,13 +1,62 @@
...
+        return sel.containsNode(n, true) ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_REJECT;
+      }
+    );
+
+    var node;
+    while (node = selectedTextNodes.nextNode()) {
+      // Work around a bug(?) in containsNode that unexpectedly returns true for
+      // the first text node in the element AFTER the selection
+      if(range.endContainer !== range.startContainer &&
+         range.endContainer.nodeType !== Node.TEXT_NODE &&
```
Comment (vanchagreen) [inline replace_text.js:48]: Do you have a way to consistently repro this? I was messing around with the script and couldn't get it to happen. I wonder if it's an interaction/bug with setting partialContainment to true in `containsNode` up above.

File: replace_text.js [hunk 2]
```diff
@@ -1,13 +1,62 @@
...
-    range = sel.getRangeAt(0);
-    range.deleteContents();
-    range.insertNode(document.createTextNode(atob(plainText)));
+  for (var r = 0; r < sel.rangeCount; ++r) {
+    var range = sel.getRangeAt(0);
+    var startOffset = range.startOffset;
+    var endOffset = range.endOffset;
+
+    var selectedTextNodes = document.createNodeIterator(
+      range.commonAncestorContainer, NodeFilter.SHOW_TEXT, function(n) {
```
Comment (vanchagreen) [inline replace_text.js:33]: :+1: I didn't know about any of this NodeFilter stuff, this is sweet.

**Review by vanchagreen — APPROVED** [review]

**Comment by vanchagreen** [comment]
This was just a quick evening project for me, so I'm shocked that you've both found and it and practically rewrote it! This is great, and it'll fix a bunch of bugs.

I'll merge it in now and hopefully get it into the chrome web store later tonight.

Thanks so much!

File: replace_text.js (hunk 1 above)
Comment (wisnij) [reply replace_text.js:48]: In particular I noticed that happening when I triple-clicked on some `<p>` text to select the entire paragraph. I suspect that internally it was counting all the elements _after_ that paragraph as part of the selection as well, all the way up until the start of the next actual text. This `if` was kind of a hacky workaround but it seemed to fix the problem at least in that specific case.

The value I gave to partialContainment didn't seem to affect the results either way in any situation that I tried, so either that isn't supported by the version of Chrome I was running, or I wasn't trying the right things. I am not primarily a web dev person so I'm not really sure what was going on here.

**Comment by wisnij** [comment]
No problem :) Your extension was the first one I found that both (a) rot13'ed text in place and (b) actually had a link to the source on github, so it was a natural place to start.

File: replace_text.js (hunk 1 above)
Comment (vanchagreen) [reply replace_text.js:48]: Yeah, it wouldn't surprise me if a triple-click created selection is a bit wonky. It's also interesting that the range's endContainer isn't a TEXT_NODE in that case. But yeah, I'm not sure either. A mystery for another day :smile:

**Comment by vanchagreen** [comment]
Just out of curiosity: a natural place to start what? :grin: No need to answer if you don't want to!

**Comment by wisnij** [comment]
I mean instead of starting completely from scratch and trying to write my own extension, which I'd never done before.

**Comment by vanchagreen** [comment]
Ahh gotcha. Well glad it helped! And I'd definitely recommend writing a chrome extension from scratch sometime; they're super fun

**Pull request merged**